src/lib/reasoners/uf.ml

```diff
@@ -999,95 +998,42 @@ let assign_next env =
   Debug.check_invariants "assign_next" env;
   res, env
 
-(**** Counter examples functions ****)
-let is_a_good_model_value (x, _) =
-  match X.leaves x with
-    [] -> true
-  | [y] -> X.equal x y
-  | _ -> false
+let save_cache () =
+  LX.save_cache ()
 
-let is_const_term (x, _) =
-  match X.term_extract x with
-  | Some t, _ ->
-    E.const_term t
-  | _ ->
-    (*cannot test for theories which don't implement term_extract*)
-    true
+let reinit_cache () =
+  LX.reinit_cache ()
+
+(****************************************************************************)
+(*                      Model generation functions                          *)
+(****************************************************************************)
 
 let model_repr_of_term t env mrepr =
   try ME.find t mrepr, mrepr
   with Not_found ->
     let mk = try ME.find t env.make with Not_found -> assert false in
-    let rep,_ = try MapX.find mk env.repr with Not_found -> assert false in
+    let rep, _ = try MapX.find mk env.repr with Not_found -> assert false in
     let cls =
-      try SE.elements (MapX.find rep env.classes)
+      try
+        Expr.Set.elements (MapX.find rep env.classes)
+        |> List.map (fun t -> Expr.Map.find t env.make)
+        |> List.filter X.is_constant
       with Not_found -> assert false
     in
-    let cls =
-      try List.rev_map (fun s -> s, ME.find s env.make) cls
-      with Not_found -> assert false
+    let tt =
+      match cls with
+      | [] ->
+        (* We do this because terms of some semantic values created by CS
+           aren't created and added to the union-find. *)
+        X.to_const_term rep
```

Comment:
Not sure what this comment means; terms of *most* semantic values created by Alt-Ergo are not added to `env.make`. I'd have expected `X.to_const_term rep` to always be the correct answer (except for arrays). When is that not the case?

(My understanding is that `cls` is guaranteed to be either `[]` or `[rep]` and any other value would break Shostak invariants; is that correct?)

Reply:
I agree. I reproduced the old code here but we don't need to filter the class at all. This code seems to work:
```ocaml 
    let mk = try ME.find t env.make with Not_found -> assert false in
    let rep, _ = try MapX.find mk env.repr with Not_found -> assert false in
    match X.to_const_term rep with 
    | Some v -> v, ME.add t v mrepr
    | None -> assert false
```
If `X.to_const_term rep` fails to produce a model term here, it's a bug.